internal/backend/runtime/omni/controllers/omni/pending_machine_status.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2025 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package omni
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic/qtransform"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/google/uuid"
+	"github.com/siderolabs/gen/xerrors"
+	"github.com/siderolabs/talos/pkg/machinery/client"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/meta"
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/siderolink"
+	"github.com/siderolabs/omni/internal/backend/runtime/talos"
+)
+
+// PendingMachineStatusController is a controller that writes the meta key to the pending machine.
+type PendingMachineStatusController = qtransform.QController[*siderolink.PendingMachine, *siderolink.PendingMachineStatus]
+
+// NewPendingMachineStatusController initializes PendingMachineStatusController.
+func NewPendingMachineStatusController() *PendingMachineStatusController {
+	return qtransform.NewQController(
+		qtransform.Settings[*siderolink.PendingMachine, *siderolink.PendingMachineStatus]{
+			Name: "PendingMachineStatusController",
+			MapMetadataFunc: func(pendingMachine *siderolink.PendingMachine) *siderolink.PendingMachineStatus {
+				return siderolink.NewPendingMachineStatus(resources.EphemeralNamespace, pendingMachine.Metadata().ID())
+			},
+			UnmapMetadataFunc: func(pendingMachineStatus *siderolink.PendingMachineStatus) *siderolink.PendingMachine {
+				return siderolink.NewPendingMachine(resources.EphemeralNamespace, pendingMachineStatus.Metadata().ID(), nil)
+			},
+			TransformFunc: func(ctx context.Context, r controller.Reader, logger *zap.Logger, pendingMachine *siderolink.PendingMachine, pendingMachineStatus *siderolink.PendingMachineStatus) error {
+				_, err := safe.ReaderGetByID[*siderolink.LinkStatus](ctx, r, pendingMachine.Metadata().ID())
+				if err != nil {
+					if state.IsNotFoundError(err) {
+						return xerrors.NewTagged[qtransform.SkipReconcileTag](err)
+					}
+
+					return err
+				}
+
+				c, err := getClient(ctx, r, pendingMachine)
+				if err != nil {
+					return err
+				}
+
+				defer c.Close() //nolint:errcheck
+
+				ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+				defer cancel()
+
+				if err = handleUUIDConflicts(ctx, r, c, logger, pendingMachine, pendingMachineStatus); err != nil {
+					return err
+				}
+
+				return generateUniqueNodeToken(ctx, c, logger, pendingMachineStatus)
+			},
+		},
+		qtransform.WithExtraMappedInput(qtransform.MapperSameID[*siderolink.LinkStatus, *siderolink.PendingMachine]()),
```

Comment:
Do we need `handleInputs` on this?

Reply:
I think we don't care about the link status removal here. Only for it to appear, so I think we can ignore that